hazelcast/src/main/java/com/hazelcast/spi/impl/merge/AbstractContainerCollector.java

```diff
@@ -0,0 +1,376 @@
+            Collection<C> containers = new ArrayList<C>();
+
+            Iterator<C> iterator = containerIterator(partitionId);
+            while (iterator.hasNext()) {
+                C container = iterator.next();
+                // just collect owned partitions
+                if (partitionService.isPartitionOwner(partitionId)) {
+                    MergePolicyConfig mergePolicyconfig = getMergePolicyConfig(container);
+                    SplitBrainMergePolicy mergePolicy = mergePolicyProvider.getMergePolicy(mergePolicyconfig.getPolicy());
+                    if (isMergeable(container) && !(mergePolicy instanceof DiscardMergePolicy)) {
+                        containers.add(container);
+                    } else {
+                        destroy(container);
+                    }
+                } else {
+                    destroyBackup(container);
+                }
+                iterator.remove();
+            }
+
+            if (!containers.isEmpty()) {
+                containersByPartitionId.put(partitionId, containers);
+            }
+        }
+    }
+
+    /**
+     * Collects a non-partitioned container with a given name from a specific partition ID.
+     */
+    private final class CollectNonPartitionedContainerRunnable implements PartitionSpecificRunnable {
+
+        private final int partitionId;
+        private final String containerName;
+
+        CollectNonPartitionedContainerRunnable(int partitionId, String containerName) {
+            this.partitionId = partitionId;
+            this.containerName = containerName;
+        }
+
+        @Override
+        public int getPartitionId() {
+            return partitionId;
+        }
+
+        @Override
+        public void run() {
+            try {
+                collectContainer();
+            } finally {
+                latch.countDown();
+            }
+        }
+
+        private void collectContainer() {
+            C container = getContainer(containerName);
+            // just collect owned partitions
+            if (partitionService.isPartitionOwner(partitionId)) {
+                MergePolicyConfig mergePolicyconfig = getMergePolicyConfig(container);
+                SplitBrainMergePolicy mergePolicy = mergePolicyProvider.getMergePolicy(mergePolicyconfig.getPolicy());
+                if (isMergeable(container) && !(mergePolicy instanceof DiscardMergePolicy)) {
+                    Collection<C> containers = getContainerCollection();
+                    containers.add(container);
+                } else {
+                    destroy(container);
+                }
+            } else {
+                destroyBackup(container);
+            }
+        }
+
+        private Collection<C> getContainerCollection() {
+            Collection<C> containers = containersByPartitionId.get(partitionId);
+            if (containers == null) {
+                Collection<C> candidate = new ConcurrentLinkedQueue<C>();
```

Comment:
We concurrently start one `CollectNonPartitionedContainerRunnable` per non-partitioned container in the data structure. The containers can all be in the same partition or in different ones. Since the final map will be a `ConcurrentMap<Integer, Collection<Container>>`, we also need a concurrent collection here. Otherwise the concurrent `Runnable`s can't populate it in parallel.